Add implicit diffusion

asalmgren opened the pull request:
Tested with diffusive CFL ~ 50. Recommend `erf.vert_implicit_fac = 1 1 0` -- see below for more discussion.

asalmgren committed:
changes to allow for implicit vertical diffusion of theta

asalmgren committed:
Merge branch 'development' into clean_up

asalmgren committed:
Merge branch 'development' into clean_up

asalmgren committed:
Merge branch 'development' into clean_up

asalmgren committed:
first pass at implicit diffusion for (rho theta)

asalmgren committed:
Merge branch 'development' into add_implicit_diffusion

asalmgren committed:
add _N version of implicit

ewquon committed:
Cleanup, reorg files, create stub for _T

asalmgren committed:
Merge branch 'development' into add_implicit_diffusion

ewquon committed:
Get rid of unused gamma, update stub

ewquon committed:
Add assert, comment out dirichlet

ewquon committed:
Use dtau instead of slow_dt

ewquon committed:
Inhomog neumann -- verified on zhi for N path

ewquon committed:
Add metric to LHS coeffs

ewquon committed:
Scratch that

ewquon committed:
Fill z_phys_nd for all zhi ghost cells

ewquon committed:
Make neumann behavior consistent on zlo/zhi

ewquon committed:
Merge branch 'development' of github.com:erf-model/ERF into add_implicit_diffusion

ewquon committed:
Rearrange input processing order

so that if grid_stretching_ratio is specified, mesh_type = StretchedDz

ewquon committed:
Avoid potentially confusing nz definition

ewquon committed:
Fix for stretched_dz_h initialization

asalmgren committed:
Merge branch 'development' into add_implicit_diffusion

ewquon committed:
Inhomog neumann fully implemented for N,S paths

verified with specified theta_grad zlo,zhi and surface_layer on zlo

ewquon committed:
Reorg input files

ewquon committed:
Add diffusion test input files

ewquon committed:
Allow control over vert_implicit_fac per stage

ewquon committed:
Document vert_implicit_fac

ewquon committed:
Cleanup

ewquon committed:
Copy ImplicitDiffForState_S to _T

ewquon committed:
Cleanup arg lists for ImplicitDiffForState_*

ewquon committed:
Fix domain boundary check

ewquon committed:
Create separate ERF_Setup*Diff.H; rename stuff (sorry)

ewquon committed:
Differentiate ERF_Setup*Diff.H headers, to address unused vars warnings

ewquon committed:
Merge branch 'development' of github.com:erf-model/ERF into add_implicit_diffusion

ewquon committed:
Add missing (1 - implicit_fac)

ewquon committed:
Use precalculated stretched_dz in ImplicitDiffForState_S

ewquon committed:
Add implicit diffusion verification notes

ewquon committed:
Make it markdown format

ewquon commented:
Verification of implicit diffusion completed for no-terrain and stretched grid cases. Details in Exec/ABL/column_diffusion_test/README.md. tl;dr below

**Test problem**: Horizontally homogeneous (periodic boundaries) with a neutrally stratified layer below a
capping inversion and a stably stratified layer aloft; diffusivity is constant throughout. Three
sets of BCs were tested:

* Neumann on zlo, specified theta_grad on zhi
* Specified theta_grad on zlo, zhi
* Surface layer on zlo, specified theta_grad on zhi

**Summary of findings**

* The original fully explicit setup requires a diffusive CFL < 0.5 for numerical stability
* No-terrain and stretched grid give the same level of error compared to explicit (in terms of MAE
  and RMSE)
* Implicit veritical allows diffusive CFL ~ 50 (dt=60 s) for this test problem, above which the
  solution remains numerically stable but error begins to increase significantly
* To retain full temporal accuracy -- and minimize error -- use explicit diffusion in the final RK
  stage (e.g., with `erf.vert_implicit_fac = 1 1 0`)

ewquon committed:
spell

ewquon committed:
Hide PBL + moisture warning

nataraj2 committed:
Adding input parameter to ERF_Datastruct.H (#2632)

* Adding parameter for time interval between data files for hindcast

* Correcitng ocmpilation error

* Removing duplicated initialization

---------

Co-authored-by: Mahesh Natarajan <nataraj2@login35.chn.perlmutter.nersc.gov>
Co-authored-by: Mahesh Natarajan <nataraj2@login11.chn.perlmutter.nersc.gov>

asalmgren committed:
replace dx(2) by dzmin in microphysics (#2630)

* replace dx(2) by dzmin in microphysics

* Pass global min of dz with terrain to micro model.

* add override function to lagrangian micro model.

* Lag is not multilevel.

---------

Co-authored-by: Aaron M. Lattanzi <103702284+AMLattanzi@users.noreply.github.com>
Co-authored-by: AMLattanzi <AMLattanzi@lbl.gov>

AMLattanzi committed:
pass dz_min with restart. (#2634)

asalmgren committed:
Some cleanup (#2633)

* some cleanup

* multiply gp by mapfac

asalmgren committed:
more cleanup plus laying groundwork (#2635)

asalmgren committed:
dont need vert ghost cells for mapfac; fix OOB in upwind_real_bcs

ewquon committed:
Fix logic for calling ImplicitDiffForState_*

ewquon committed:
Pass neumann BC vals to ImplicitDiffForState_*

ewquon committed:
Fix sign on imposed theta_grad on zlo

ewquon committed:
SurfLayer takes precedence over neumann on zlo; cleanup

ewquon committed:
Cleanup more unused params

ewquon committed:
More unused vars

ewquon committed:
Merge branch 'development' of github.com:erf-model/ERF into add_implicit_diffusion

ewquon committed:
Fix get_fft_bc call

ewquon commented:
More verification notes (to self) for the column problem defined above. With the _same time step_:

* Explicit vs imex (`erf.vert_implicit_fac = 1 1 0`) have an **abs err in theta ~O(1e-7)**
* This holds for the column problem defined above for free-slip (`erf.surface_layer.flux_type = "custom"` and `erf.most.ustar = 0`) as well as with shear stress (`wrf.most.ustar = 0.5`)
* This holds for constant diffusivity (`erf.molec_diff_type = "Constant"`) and height-varying diffusivity (`erf.molec_diff_type = "ConstantAlpha"`, where K = rho(z)*alpha)
* This does NOT hold for `erf.molec_diff_type = "None"` and `erf.pbl_type = "MRF"`.  Abs err in theta ~O(1e-3). Suspect this is not specific to MRF.

asalmgren committed:
update amrex submodule (#2637)

pressel committed:
Morrison Microphysics: Add Plotting Capabilities and Configurable Droplet Concentrations (#2636)

* Morrison microphysics: Add plotting capabilities and configurable droplet concentration

This commit adds new functionality to the Morrison two-moment microphysics scheme:

1. **Plotting Infrastructure (ERF_Morrison_Plot.cpp)**
   - Added GetPlotVarNames() and GetPlotVar() methods to Morrison class
   - Enables output of 19 Morrison microphysics variables to plotfiles
   - Includes: density, temperature, pressure, moisture variables (qv, qc, qi, qn, qt, qp)
     and hydrometeor variables (qrain, qsnow, qgraup, nc, nr, ni, ns, ng, omega)
   - Provides diagnostic capability for detailed analysis of Morrison scheme behavior

2. **Configurable Droplet Concentration**
   - Added runtime configuration of constant droplet concentration (NDCNST)
   - New input parameter: erf.morrison_ndcnst (default value preserved)
   - Fortran interface updated with set_morrison_ndcnst_c() function
   - Allows customization of cloud droplet number concentration for different cases
   - Applied to both C++ and Fortran implementations

3. **Build System Updates**
   - Added ERF_Morrison_Plot.cpp to CMake build configuration
   - Updated Make.package with new source file

Files changed:
- Source/Microphysics/Morrison/ERF_Morrison.H: Added plotting method declarations
- Source/Microphysics/Morrison/ERF_AdvanceMorrison.cpp: Added NDCNST configuration
- Source/Microphysics/Morrison/ERF_Morrison_Fortran_Interface.H: Added Fortran interface
- Source/Microphysics/Morrison/ERF_Morrison_Plot.cpp: New plotting implementation
- Source/Microphysics/Morrison/ERF_module_mp_morr_two_moment.F90: Added SET_MORRISON_NDCNST
- Source/Microphysics/Morrison/Make.package: Added plot source
- CMake/BuildERFExe.cmake: Added plot source to build

These changes enhance the Morrison microphysics module with improved diagnostic
capabilities and greater flexibility for research applications.

* Documentation.

* Documentation.

* Remove debug print statement from Morrison NDCNST configuration

Remove the print statement that was outputting NDCNST value during
runtime configuration. This reduces unnecessary console output during
normal simulation runs.

asalmgren committed:
Fix spell (#2638)

* update amrex submodule

* fix spelling

akashdhruv committed:
Add CPU-GPU data copy and synchronization call to Noah-MP interface (#2613)

* Initial GPU interface

* bug fix in indexing

* Merge with upstream and fix trailing whitespace errors

* Fix again

* Bug fix in CPU GPU loops

---------

Co-authored-by: Akash <adhruv@polaris-login-04.hsn.cm.polaris.alcf.anl.gov>

skang67 committed:
EB: Enable TwoWay coupling for multilevel EB. (#2640)

* Set afrea fractions at the regular and covered box boundaries.

* Correct cell flags for zero volume fractions.

* Set flag covered if volume fraction is zero.

* Set disconnected and covered for zero-volume cells in eb_aux_.

* set_connection_flags on bx_grown.

* set_connection_flags over trimmed grown box to avoid out-of-bound error.

* Removed if conditions for isSingleValued() checks.

* Use std::abs to check small values for aux_vcent and aux_bcent.

* High-order approximation of pressure gradient.

* Added EB_SquareCylinder to CMakeLists.

* Debug: Fix EB_average_down arguments.

* Set cell_vol to 1.0.

* Set the minimum upwind order to CEN2.

* Commented unused variable.

* Removed whitespaces.

---------

Co-authored-by: Soonpil Kang <kang18@llnl.gov>

ewquon committed:
Pass implicit_fac to ComputeStrain_*, update FAB instantiation logic

asalmgren committed:
remove unused gOcp (#2641)

jmsexton03 committed:
Add Perlmutter cmake script: GTL CUDA + 6 pkgs (#2644)

* Add Perlmutter cmake script: GTL CUDA + 6 pkgs

- Add cmake build script for NERSC Perlmutter with GPU support
- Link mpi_gtl_cuda library for GPU-aware MPI communication
- Configure for NVIDIA A100 (Ampere 80) architecture
- Enable Kokkos, EKAT, RRTMGP, FFT, NetCDF, and HDF5

* Add second serial call to make

* Order for second call was wrong

* Add no-as-needed, update to newer modules

* filter out as-needed

ewquon committed:
Move ComputeSmnSmn into ComputeStrain_*

All ComputeStrain functions include all strain FABs

AMLattanzi committed:
export sycl for kokkos. (#2645)

ewquon committed:
Properly allocate tau with implicit and no terrain; don't need tau21

akashdhruv committed:
Cleaning up plotting logic for Noah-MP and adding files for a regression test (#2643)

* A proper way to plot land files

* add wps test files for noahmp

* Update README to provide instructions for noahmp test

---------

Co-authored-by: Akash <adhruv@polaris-login-02.hsn.cm.polaris.alcf.anl.gov>
Co-authored-by: Akash <adhruv@polaris-login-04.hsn.cm.polaris.alcf.anl.gov>

ewquon committed:
Merge branch 'development' of github.com:erf-model/ERF into add_implicit_diffusion

ewquon committed:
git pull didn't merge as expected?

ewquon commented:
Implicit rho*theta diffusion should be ready for testing at this point. Refactor of strain and stress calcs is underway to make tau31 and tau32 semi-implicit.
